File: core/src/main/resources/org/treblereel/gwt/three4g/materials/Material.js
import { EventDispatcher } from '../core/EventDispatcher.js';
import { FrontSide, FlatShading, NormalBlending, LessEqualDepth, AddEquation, OneMinusSrcAlphaFactor, SrcAlphaFactor, AlwaysStencilFunc, KeepStencilOp } from '../constants.js';
import { MathUtils } from '../math/MathUtils.js';
import { Vector2 } from '../math/Vector2.js';

import { Color } from './../math/Color.js';
import { MaterialParameters } from '../closure/material/MaterialParameters.js';
import { MaterialInterface } from '../closure/material/MaterialInterface.js';

/** @type {number} */
let materialId = 0;

class Material extends EventDispatcher {

	constructor() {

		super();

	//Object.defineProperty( this, 'id', { value: materialId ++ } );

		/** @type {number} */
		this.id =  materialId ++;
		/** @type {string} */
		this.uuid = MathUtils.generateUUID();

		/** @type {string} */
		this.name = '';
		/** @type {string} */
		this.type = 'Material';

		/** @type {boolean} */
		this.fog = true;
		/** @type {number} */
		this.blending = NormalBlending;
		/** @type {number} */
		this.side = FrontSide;
		/** @type {boolean} */
		this.vertexColors = false;
		/** @type {number} */
		this.opacity = 1;
		/** @type {boolean} */
		this.transparent = false;
		/** @type {number} */
		this.blendSrc = SrcAlphaFactor;
		/** @type {number} */
		this.blendDst = OneMinusSrcAlphaFactor;
		/** @type {number} */
		this.blendEquation = AddEquation;
		/** @type {number|null} */
		this.blendSrcAlpha = null;
		/** @type {number|null} */
		this.blendDstAlpha = null;
		/** @type {number|null} */
		this.blendEquationAlpha = null;
		/** @type {number} */
		this.depthFunc = LessEqualDepth;
		/** @type {boolean} */
		this.depthTest = true;
		/** @type {boolean} */
		this.depthWrite = true;
		/** @type {number} */
		this.stencilWriteMask = 0xff;
		/** @type {number} */
		this.stencilFunc = AlwaysStencilFunc;
		/** @type {number} */
		this.stencilRef = 0;
		/** @type {number} */
		this.stencilFuncMask = 0xff;
		/** @type {number} */
		this.stencilFail = KeepStencilOp;
		/** @type {number} */
		this.stencilZFail = KeepStencilOp;
		/** @type {number} */
		this.stencilZPass = KeepStencilOp;
		/** @type {boolean} */
		this.stencilWrite = false;

		this.clippingPlanes = null;
		/** @type {boolean} */
		this.clipIntersection = false;
		/** @type {boolean} */
		this.clipShadows = false;

		this.shadowSide = null;
		/** @type {boolean} */
		this.colorWrite = true;

		this.precision = null; // override the renderer's default precision for this material
		/** @type {boolean} */
		this.polygonOffset = false;
		/** @type {number} */
		this.polygonOffsetFactor = 0;
		/** @type {number} */
		this.polygonOffsetUnits = 0;
		/** @type {boolean} */
		this.dithering = false;

		/** @type {number} */
		this.alphaTest = 0;
		/** @type {boolean} */
		this.premultipliedAlpha = false;
		/** @type {boolean} */
		this.visible = true;
		/** @type {boolean} */
		this.toneMapped = true;

		this.userData = {};

		/** @type {number} */
		this.version = 0;
		/** @type {boolean} */
		this.isMaterial = true;
		/** @type {boolean|undefined} */
		this.flatShading;

	}

	onBeforeCompile( /* shaderobject, renderer */ ) {}

	customProgramCacheKey() {

		return this.onBeforeCompile.toString();

	}

	/**
	 * 
	 * @param {Object=} values 
	 */
	setValues( values ) {

		if ( values === undefined ) return;

		for ( const key in values ) {

			const /** @type {number} */ newValue = values[ key ];

			if ( newValue === undefined ) {

				console.warn( 'THREE.Material: \'' + key + '\' parameter is undefined.' );
				continue;

			}

			// for backward compatability if shading is set in the constructor
			if ( key === 'shading' ) {

				console.warn( 'THREE.' + this.type + ': .shading has been removed. Use the boolean .flatShading instead.' );
				this.flatShading = ( newValue === FlatShading ) ? true : false;
				continue;

			}

			/** @type {*} */
			const currentValue = this[ key ];

			if ( currentValue === undefined ) {

				console.warn( 'THREE.' + this.type + ': \'' + key + '\' is not a property of this material.' );
				continue;

			}

			if ( currentValue && currentValue.isColor ) {

				currentValue.set( newValue );

			} else if ( ( currentValue && currentValue.isVector3 ) && ( newValue && newValue.isVector3 ) ) {

				currentValue.copy( newValue );

			} else {

				this[ key ] = newValue;

			}

		}

	}

	/**
	 * //TODO
	* @param {Object=} meta
	* @suppress{checkTypes}
	*/
	toJSON( meta ) {

		const isRoot = ( meta === undefined || typeof meta === 'string' );

		if ( isRoot ) {

			meta = {
				textures: {},
				images: {}
			};

		}

		const data = {
			metadata: {
				version: 4.5,
				type: 'Material',
				generator: 'Material.toJSON'
			}
		};

		/** @type {MaterialInterface} */
		var that = this;

		// standard Material serialization
		data.uuid = that.uuid;
		data.type = that.type;

		if ( that.name !== '' ) data.name = that.name;

		if ( that.color && that.color.isColor ) data.color = that.color.getHex();

		if ( that.roughness !== undefined ) data.roughness = that.roughness;
		if ( that.metalness !== undefined ) data.metalness = that.metalness;

		if ( that.sheen && that.sheen.isColor ) data.sheen = that.sheen.getHex();
		if ( that.emissive && that.emissive.isColor ) data.emissive = that.emissive.getHex();
		if ( that.emissiveIntensity && that.emissiveIntensity !== 1 ) data.emissiveIntensity = that.emissiveIntensity;

		if ( that.specular && that.specular.isColor ) data.specular = that.specular.getHex();
		if ( that.shininess !== undefined ) data.shininess = that.shininess;
		if ( that.clearcoat !== undefined ) data.clearcoat = that.clearcoat;
		if ( that.clearcoatRoughness !== undefined ) data.clearcoatRoughness = that.clearcoatRoughness;

		if ( that.clearcoatMap && that.clearcoatMap.isTexture ) {

			data.clearcoatMap = that.clearcoatMap.toJSON( meta ).uuid;

		}

		if ( that.clearcoatRoughnessMap && that.clearcoatRoughnessMap.isTexture ) {

			data.clearcoatRoughnessMap = that.clearcoatRoughnessMap.toJSON( meta ).uuid;

		}

		if ( that.clearcoatNormalMap && that.clearcoatNormalMap.isTexture ) {

			data.clearcoatNormalMap = that.clearcoatNormalMap.toJSON( meta ).uuid;
			data.clearcoatNormalScale = that.clearcoatNormalScale.toArray();

		}

		if ( that.map && that.map.isTexture ) data.map = that.map.toJSON( meta ).uuid;
		if ( that.matcap && that.matcap.isTexture ) data.matcap = that.matcap.toJSON( meta ).uuid;
		if ( that.alphaMap && that.alphaMap.isTexture ) data.alphaMap = that.alphaMap.toJSON( meta ).uuid;

		if ( that.lightMap && that.lightMap.isTexture ) {

			data.lightMap = that.lightMap.toJSON( meta ).uuid;
			data.lightMapIntensity = that.lightMapIntensity;

		}

		if ( that.aoMap && that.aoMap.isTexture ) {

			data.aoMap = that.aoMap.toJSON( meta ).uuid;
			data.aoMapIntensity = that.aoMapIntensity;

		}

		if ( that.bumpMap && that.bumpMap.isTexture ) {

			data.bumpMap = that.bumpMap.toJSON( meta ).uuid;
			data.bumpScale = that.bumpScale;

		}

		if ( that.normalMap && that.normalMap.isTexture ) {

			data.normalMap = that.normalMap.toJSON( meta ).uuid;
			data.normalMapType = that.normalMapType;
			data.normalScale = that.normalScale.toArray();

		}

		if ( that.displacementMap && that.displacementMap.isTexture ) {

			data.displacementMap = that.displacementMap.toJSON( meta ).uuid;
			data.displacementScale = that.displacementScale;
			data.displacementBias = that.displacementBias;

		}

		if ( that.roughnessMap && that.roughnessMap.isTexture ) data.roughnessMap = that.roughnessMap.toJSON( meta ).uuid;
		if ( that.metalnessMap && that.metalnessMap.isTexture ) data.metalnessMap = that.metalnessMap.toJSON( meta ).uuid;

		if ( that.emissiveMap && that.emissiveMap.isTexture ) data.emissiveMap = that.emissiveMap.toJSON( meta ).uuid;
		if ( that.specularMap && that.specularMap.isTexture ) data.specularMap = that.specularMap.toJSON( meta ).uuid;

		if ( that.envMap && that.envMap.isTexture ) {

			data.envMap = that.envMap.toJSON( meta ).uuid;
			data.reflectivity = that.reflectivity; // Scale behind envMap
			data.refractionRatio = that.refractionRatio;

			if ( that.combine !== undefined ) data.combine = that.combine;
			if ( that.envMapIntensity !== undefined ) data.envMapIntensity = that.envMapIntensity;

		}

		if ( that.gradientMap && that.gradientMap.isTexture ) {

			data.gradientMap = that.gradientMap.toJSON( meta ).uuid;

		}

		if ( that.size !== undefined ) data.size = that.size;
		if ( that.sizeAttenuation !== undefined ) data.sizeAttenuation = that.sizeAttenuation;

		if ( that.blending !== NormalBlending ) data.blending = that.blending;
		if ( that.side !== FrontSide ) data.side = that.side;
		if ( that.vertexColors ) data.vertexColors = true;

		if ( that.opacity < 1 ) data.opacity = that.opacity;
		if ( that.transparent === true ) data.transparent = that.transparent;

		data.depthFunc = that.depthFunc;
		data.depthTest = that.depthTest;
		data.depthWrite = that.depthWrite;

		data.stencilWrite = that.stencilWrite;
		data.stencilWriteMask = that.stencilWriteMask;
		data.stencilFunc = that.stencilFunc;
		data.stencilRef = that.stencilRef;
		data.stencilFuncMask = that.stencilFuncMask;
		data.stencilFail = that.stencilFail;
		data.stencilZFail = that.stencilZFail;
		data.stencilZPass = that.stencilZPass;

		// rotation (SpriteMaterial)
		if ( that.rotation && that.rotation !== 0 ) data.rotation = that.rotation;

		if ( that.polygonOffset === true ) data.polygonOffset = true;
		if ( that.polygonOffsetFactor !== 0 ) data.polygonOffsetFactor = that.polygonOffsetFactor;
		if ( that.polygonOffsetUnits !== 0 ) data.polygonOffsetUnits = that.polygonOffsetUnits;

		if ( that.linewidth && that.linewidth !== 1 ) data.linewidth = that.linewidth;
		if ( that.dashSize !== undefined ) data.dashSize = that.dashSize;
		if ( that.gapSize !== undefined ) data.gapSize = that.gapSize;
		if ( that.scale !== undefined ) data.scale = that.scale;

		if ( that.dithering === true ) data.dithering = true;

		if ( that.alphaTest > 0 ) data.alphaTest = that.alphaTest;
		if ( that.premultipliedAlpha === true ) data.premultipliedAlpha = that.premultipliedAlpha;

		if ( that.wireframe === true ) data.wireframe = that.wireframe;
		if ( that.wireframeLinewidth > 1 ) data.wireframeLinewidth = that.wireframeLinewidth;
		if ( that.wireframeLinecap !== 'round' ) data.wireframeLinecap = that.wireframeLinecap;
		if ( that.wireframeLinejoin !== 'round' ) data.wireframeLinejoin = that.wireframeLinejoin;

		if ( that.morphTargets === true ) data.morphTargets = true;
		if ( that.morphNormals === true ) data.morphNormals = true;
		if ( that.skinning === true ) data.skinning = true;

		if ( that.flatShading === true ) data.flatShading = that.flatShading;

		if ( that.visible === false ) data.visible = false;

		if ( that.toneMapped === false ) data.toneMapped = false;

		if ( JSON.stringify( that.userData ) !== '{}' ) data.userData = that.userData;

		// TODO: Copied from Object3D.toJSON

		function extractFromCache( cache ) {

			const values = [];

			for ( const key in cache ) {

				const data = cache[ key ];
				delete data.metadata;
				values.push( data );

			}

			return values;

		}

		if ( isRoot ) {

			const textures = extractFromCache( meta.textures );
			const images = extractFromCache( meta.images );

			if ( textures.length > 0 ) data.textures = textures;
			if ( images.length > 0 ) data.images = images;

		}

		return data;

	}

	/**
	 * @return {Material} 
	 */
	clone() {

		return new this.constructor().copy( this );

	}

	/**
	 * @param {Material} source
	 * @return {Material} 
	 */
	copy( source ) {

		this.name = source.name;

		this.fog = source.fog;

		this.blending = source.blending;
		this.side = source.side;
		this.vertexColors = source.vertexColors;

		this.opacity = source.opacity;
		this.transparent = source.transparent;

		this.blendSrc = source.blendSrc;
		this.blendDst = source.blendDst;
		this.blendEquation = source.blendEquation;
		this.blendSrcAlpha = source.blendSrcAlpha;
		this.blendDstAlpha = source.blendDstAlpha;
		this.blendEquationAlpha = source.blendEquationAlpha;

		this.depthFunc = source.depthFunc;
		this.depthTest = source.depthTest;
		this.depthWrite = source.depthWrite;

		this.stencilWriteMask = source.stencilWriteMask;
		this.stencilFunc = source.stencilFunc;
		this.stencilRef = source.stencilRef;
		this.stencilFuncMask = source.stencilFuncMask;
		this.stencilFail = source.stencilFail;
		this.stencilZFail = source.stencilZFail;
		this.stencilZPass = source.stencilZPass;
		this.stencilWrite = source.stencilWrite;

		const srcPlanes = source.clippingPlanes;
		let dstPlanes = null;

		if ( srcPlanes !== null ) {

			const n = srcPlanes.length;
			dstPlanes = new Array( n );

			for ( let i = 0; i !== n; ++ i ) {

				dstPlanes[ i ] = srcPlanes[ i ].clone();

			}

		}

		this.clippingPlanes = dstPlanes;
		this.clipIntersection = source.clipIntersection;
		this.clipShadows = source.clipShadows;

		this.shadowSide = source.shadowSide;

		this.colorWrite = source.colorWrite;

		this.precision = source.precision;

		this.polygonOffset = source.polygonOffset;
		this.polygonOffsetFactor = source.polygonOffsetFactor;
		this.polygonOffsetUnits = source.polygonOffsetUnits;

		this.dithering = source.dithering;

		this.alphaTest = source.alphaTest;
		this.premultipliedAlpha = source.premultipliedAlpha;

		this.visible = source.visible;

		this.toneMapped = source.toneMapped;

		this.userData = JSON.parse( JSON.stringify( source.userData ) );

		return this;

	}

	dispose() {

		this.dispatchEvent( { type: 'dispose' } );

	}

}

Object.defineProperty( Material.prototype, 'needsUpdate', {

	set( value ) {

		if ( value === true ) this.version ++;

	}

} );

export { Material };

File: core/src/main/java/org/treblereel/gwt/three4g/materials/Material.java
package org.treblereel.gwt.three4g.materials;

import jsinterop.annotations.JsType;
import org.treblereel.gwt.three4g.core.PropertyHolder;
import org.treblereel.gwt.three4g.core.events.EventDispatcher;
import org.treblereel.gwt.three4g.materials.parameters.MaterialParameters;
import org.treblereel.gwt.three4g.math.Plane;

/**
 * @author Dmitrii Tikhomirov
 * Created by treblereel on 2/27/18.
 */
@JsType(isNative = true, namespace = "THREE")
public class Material extends EventDispatcher {

    /**
     * Sets the alpha value to be used when running an alpha test. The material will not be renderered if the opacity is lower than this value. Default is 0.
     */
    public float alphaTest;

    /**
     * Blending destination. Default is OneMinusSrcAlphaFactor. See the destination factors constants for all possible values.
     * The material's # .blending must be set to CustomBlending for this to have any effect.
     */
    public int blendDst;

    /**
     * The tranparency of the .blendDst. Default is null.
     */
    public int blendDstAlpha;

    /**
     * Blending equation to use when applying blending. Default is AddEquation. See the blending equation constants for all possible values.
     * The material's # .blending must be set to CustomBlending for this to have any effect.
     */
    public int blendEquation;

    /**
     * The tranparency of the .blendEquation. Default is null.
     */
    public int blendEquationAlpha;

    /**
     * Which blending to use when displaying objects with this material.
     * This must be set to CustomBlending to use custom # .blendSrc , # .blendDst or # .blendEquation .
     * See the blending mode constants for all possible values. Default is NormalBlending.
     */
    public int blending;

    /**
     * Blending source. Default is SrcAlphaFactor. See the source factors constants for all possible values.
     * The material's # .blending must be set to CustomBlending for this to have any effect.
     */
    public int blendSrc;

    /**
     * The tranparency of the .blendSrc. Default is null.
     */
    public int blendSrcAlpha;

    /**
     * Changes the behavior of clipping planes so that only their intersection is clipped, rather than their union. Default is false.
     */
    public boolean clipIntersection;

    /**
     * User-defined clipping planes specified as THREE.Plane objects in world space. These planes apply to the objects this material is attached to. Points in space whose signed distance to the plane is negative are clipped (not rendered). See the WebGL / clipping /intersection example. Default is null.
     */
    public Plane[] clippingPlanes;

    /**
     * Defines whether to clip shadows according to the clipping planes specified on this material. Default is false.
     */
    public boolean clipShadows;

    /**
     * Whether to render the material's color. This can be used in conjunction with a mesh's # .renderOrder property to create invisible objects that occlude other objects. Default is true.
     */
    public boolean colorWrite;

    /**
     * Custom depth material to be used by this material when rendering to the depth map. When shadow-casting with
     * a DirectionalLight or SpotLight, if you are (a) modifying vertex positions in the vertex shader, (b) using
     * a displacement map, (c) using an alpha map with alphaTest, or (d) using a transparent texture with alphaTest,
     * you must specify a customDepthMaterial for proper shadows. Default is undefined.
     */
    public Material customDepthMaterial;

    /**
     * Same as customDepthMaterial, but used with PointLight. Default is undefined.
     */
    public Material customDistanceMaterial;

    /**
     * Custom defines to be injected into the shader. These are passed in form of an object literal, with key/value pairs. { MY_CUSTOM_DEFINE: '' , PI2: Math.PI * 2 }. The pairs are defined in both vertex and fragment shaders. Default is undefined.
     */
    public PropertyHolder defines;


    /**
     * Which depth function to use. Default is LessEqualDepth. See the depth mode constants for all possible values.
     */
    public int depthFunc;

    /**
     * Whether to have depth test enabled when rendering this material. Default is true.
     */
    public boolean depthTest;

    /**
     * Whether rendering this material has any effect on the depth buffer. Default is true.
     * <p>
     * When drawing 2D overlays it can be useful to disable the depth writing in order to layer several things together without creating z-index artifacts.
     */
    public boolean depthWrite;

    /**
     * Whether stencil operations are performed against the stencil buffer. In order to perform writes or comparisons against the stencil buffer this value must be true. Default is false.
     */
    public boolean stencilWrite;

    /**
     * The bit mask to use when writing to the stencil buffer. Default is 0xFF.
     */
    public int stencilWriteMask;

    /**
     * The stencil comparison function to use. Default is AlwaysStencilFunc. See stencil function constants for all possible values.
     */
    public int stencilFunc;

    /**
     * The bit mask to use when comparing against the stencil buffer. Default is 0xFF.
     */
    public int stencilFuncMask;

    /**
     * The value to use when performing stencil comparisons or stencil operations. Default is 0.
     */
    public int stencilRef;

    /**
     * Which stencil operation to perform when the comparison function returns false. Default is KeepStencilOp. See the stencil operations constants for all possible values.
     */
    public int stencilFail;

    /**
     * Which stencil operation to perform when the comparison function returns true but the depth test fails. Default is KeepStencilOp. See the stencil operations constants for all possible values.
     */
    public int stencilZFail;

    /**
     * Which stencil operation to perform when the comparison function returns true and the depth test passes. Default is KeepStencilOp. See the stencil operations constants for all possible values.
     */
    public int stencilZPass;

    /**
     * Define whether the material is rendered with flat shading. Default is false.
     */
    public boolean flatShading;

    /**
     * Whether the material is affected by fog. Default is true.
     */
    public boolean fog;

    /**
     * Unique number for this material instance.
     */
    public int id;

    /**
     * Used to check whether this or derived classes are materials. Default is true.
     * <p>
     * You should not change this, as it used internally for optimisation.
     */
    public boolean isMaterial;

    /**
     * Whether the material is affected by lights. Default is true.
     */
    public boolean lights;


    /**
     * Optional name of the object (doesn't need to be unique). Default is an empty string.
     */
    public String name;

    /**
     * Specifies that the material needs to be updated at the WebGL level. Set it to true if you made changes that need to be reflected in WebGL.
     * This property is automatically set to true when instancing a new material.
     */
    public boolean needsUpdate;

    /**
     * Float in the range of 0.0 - 1.0 indicating how transparent the material is. A value of 0.0 indicates fully transparent, 1.0 is fully opaque.
     * If the material's # .transparent property is not set to true, the material will remain fully opaque and this value will only affect its color.
     * Default is 1.0.
     */
    public float opacity;

    /**
     * Whether to use polygon offset. Default is false. This corresponds to the GL_POLYGON_OFFSET_FILL WebGL feature.
     */
    public boolean polygonOffset;

    /**
     * Sets the polygon offset factor. Default is 0.
     */
    public int polygonOffsetFactor;

    /**
     * Sets the polygon offset units. Default is 0.
     */
    public int polygonOffsetUnits;

    /**
     * Override the renderer's default precision for this material. Can be "highp", "mediump" or "lowp". Default is null.
     */
    public String precision;

    /**
     * Whether to premultiply the alpha (transparency) value. See WebGL / Materials / Transparency for an example of the difference. Default is false.
     */
    public boolean premultipliedAlpha;

    /**
     * Whether to apply dithering to the color to remove the appearance of banding. Default is false.
     */
    public boolean dithering;

    /**
     * Defines which side of faces cast shadows. When set, can be FrontSide, BackSide, or DoubleSide. Default is null.
     * If null, the side casting shadows is determined as follows:
     * Material.side            Side casting shadows
     * FrontSide                 back side
     * BackSide                  front side
     * DoubleSide                both sides
     */
    public int shadowSide;

    /**
     * Defines which side of faces will be rendered - front, back or both. Default is FrontSide. Other options are BackSide and DoubleSide.
     */
    public int side;


    /**
     * Defines whether this material is transparent. This has an effect on rendering as transparent objects need special treatment and are rendered after non-transparent objects.
     * When set to true, the extent to which the material is transparent is controlled by setting it's # .opacity property.
     * Default is false.
     */
    public boolean transparent;


    /**
     * Value is the string 'Material'. This shouldn't be changed, and can be used to find all objects of this type in a scene.
     */
    public String type;

    /**
     * UUID of this material instance. This gets automatically assigned, so this shouldn't be edited.
     */
    public String uuid;

    /**
     * Defines whether vertex coloring is used. Default is NoColors. Other options are VertexColors and FaceColors.
     */
    public int vertexColors;

    /**
     * Defines whether precomputed vertex tangents are used. Default is false.
     */
    public boolean vertexTangents;

    /**
     * Defines whether this material is visible. Default is true.
     */
    public boolean visible;


    /**
     * An object that can be used to store custom data about the Material. It should not hold references to functions as these will not be cloned.
     */
    public PropertyHolder userData;

    public Material() {

    }

    public Material(MaterialParameters materialParameters) {

    }

    /**
     * @return a new material with the same parameters as this material.
     */
    public native Material clone();

    /**
     * Copy the parameters from the passed material into this material.
     *
     * @param material instance of Material
     * @return instance of Material
     */
    public native Material copy(Material material);

    /**
     * This disposes the material. Textures of a material don't get disposed. These needs to be disposed by Texture.
     */
    public native void dispose();

    /**
     * Sets the properties based on the values.
     *
     * @param values -- a container with parameters.
     */
    public native void setValues(Object values);

    /**
     * Convert the material to three.js JSON format.
     *
     * @return String
     */
    public native String toJSON();

    /**
     * Convert the material to three.js JSON format.
     *
     * @param meta -- object containing metadata such as textures or images for the material.
     * @return String
     */
    public native String toJSON(String meta);

    /**
     * Call # .dispatchEvent ( { type: 'update' }) on the material.
     */
    public native void update();

}
